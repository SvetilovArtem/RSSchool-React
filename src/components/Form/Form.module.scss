.form {
    border-radius: 9px;
    background-color: #21213a;
    max-width: 500px;
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding: 30px 20px;
    gap: 20px;
    color: #ffffff;
    background-repeat: no-repeat;
    margin: 50px auto;
    .label {
        margin-bottom: 8px;
    }
    .error {
        outline: 2px solid red
    }
}
.agreeBlock {
    display: flex;
    align-items: center;
    .checkbox {
        position: absolute;
        z-index: -1;
        opacity: 0;
    }
    .agree {
        cursor: pointer;
        font-size: 10px;
        display: flex;
        align-items: center;
    }
    .checkbox+label::before {
        content: '';
        display: inline-block;
        width: 20px;
        height: 20px;
        flex-shrink: 0;
        flex-grow: 0;
        border: 1px solid #adb5bd;
        border-radius: 4px;
        margin-right: 0.5em;
        background-repeat: no-repeat;
        background-position: center center;
        background-size: 70% 70%;
    }
    .checkbox:checked+label::before {
        background-image: url('../../icons/ok-icon.svg');
        background-repeat: no-repeat;
    }
}

.inputBlock, .selectBlock {
    display: flex;
    flex-direction: column;
    max-width: 280px;
    input, select {
        border: none;
        border-radius: 9px;
        outline: none;
        padding: 8px 6px;
        color: #333;
    }
}
.switcherBlock {
    display: flex;
    align-items: center;
    gap: 12px;
}
.switcher {
    display: inline-block;
    position: relative;
    width: 60px;
    height: 34px;
    input {
        display: none;
    }
    .slider {
        position: absolute;
        cursor: pointer;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
        background-color: #b16a9a;
        -webkit-transition: .4s;
        transition: .4s;
        border-radius: 34px;
    }
    .slider:before {
        position: absolute;
        content: "";
        height: 26px;
        width: 26px;
        left: 4px;
        bottom: 4px;
        background-color: white;
        -webkit-transition: .4s;
        transition: .4s;
        border-radius: 50%;
      }
      input:checked + .slider {
        background-color: #5159b6;
      }
      
      input:focus + .slider {
        box-shadow: 0 0 1px #5159b6;
      }
      
      input:checked + .slider:before {
        -webkit-transform: translateX(26px);
        -ms-transform: translateX(26px);
        transform: translateX(26px);
      }
}

.inputFileBlock {
    max-width: 280px;
    position: relative;
    margin: 15px 0;
    transition: all .3s ease;

    &:hover {
        -webkit-box-shadow: 0px 5px 10px 5px rgba(0, 145, 255, 0.471);
        -moz-box-shadow: 0px 5px 10px 5px rgba(0, 145, 255, 0.471);
        box-shadow: 0px 5px 10px 5px rgba(0, 145, 255, 0.471);
    }
    .okIcon {
        position: absolute;
        top: calc(50% - 24px);
        right: -50px;
    }
  }
   
  .inputFile {
    opacity: 0;
    visibility: hidden;
    position: absolute;
  }
   
  .inputFileIconWrapper {
    height: 60px;
    width: 60px;
    margin-right: 15px;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-align: center;
        -ms-flex-align: center;
            align-items: center;
    -webkit-box-pack: center;
        -ms-flex-pack: center;
            justify-content: center;
    border-right: 1px solid #fff;
  }
   
  .inputFileButtonText {
    line-height: 1;
    margin-top: 1px;
  }
   
  .inputFileButton {
    width: 100%;
    max-width: 290px;
    height: 60px;
    background: #1bbc9b;
    color: #fff;
    font-size: 1.125rem;
    font-weight: 700;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-align: center;
        -ms-flex-align: center;
            align-items: center;
    -webkit-box-pack: start;
        -ms-flex-pack: start;
            justify-content: flex-start;
    border-radius: 3px;
    cursor: pointer;
    // margin: 0 auto;
  }

.submitButton {
    padding: 10px 5px;
    border-radius: 9px;
    border: none;
    background: #1bbc9b;
    color: #fff;
    cursor: pointer;
    max-width: 280px;
    transition: all .3s ease;
    font-weight: 700;
    font-size: 1.125rem;
    &:hover {
        -webkit-box-shadow: 0px 5px 10px 5px rgba(0, 145, 255, 0.471);
        -moz-box-shadow: 0px 5px 10px 5px rgba(0, 145, 255, 0.471);
        box-shadow: 0px 5px 10px 5px rgba(0, 145, 255, 0.471);
    }
    &:active {
        background: #0c9b7f;
    }
}
.disabledButton {
    padding: 10px 5px;
    border-radius: 9px;
    border: none;
    background: #e5e5e5;
    color: #fff;
    cursor: pointer;
    max-width: 280px;
    transition: all .3s ease;
    font-weight: 700;
    font-size: 1.125rem;
}