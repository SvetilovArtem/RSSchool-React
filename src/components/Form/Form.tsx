import React from "react";
import { UserType } from "../../types/UserType";
import styles from "./Form.module.scss";

interface FormProps {
  setUsers: (data: UserType) => void;
}

export default class Form extends React.Component<FormProps> {
  inputName = React.createRef<HTMLInputElement>();
  inputDate = React.createRef<HTMLInputElement>();
  selectRef = React.createRef<HTMLSelectElement>();
  switcherRef = React.createRef<HTMLInputElement>();
  agreeRef = React.createRef<HTMLInputElement>();
  inputFile = React.createRef<HTMLInputElement>();
  form = React.createRef<HTMLFormElement>();
  constructor(props: FormProps) {
    super(props);
    this.onSaveUserData = this.onSaveUserData.bind(this);
  }
  state = {
    errors: {
      inputName: "",
      inputDate: "",
      selectRef: "",
    },
    isValid: false,
    downloadImg: false,
    avatar: "",
  };
  onSaveUserData = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    this.validateInput(this.inputName.current);
    this.validateInput(this.inputFile.current);
    this.validateInput(this.inputDate.current);
    this.validateInput(this.selectRef.current);

    const user: UserType = {
      name: this.inputName.current?.value || "",
      date: this.inputDate.current?.value || "",
      nationality: this.selectRef.current?.value || "",
      sex: this.switcherRef.current?.checked ? "male" : "female",
      agree: this.agreeRef.current?.checked || false,
      img: this.state.avatar || "",
    };
    this.props.setUsers(user);
  };
  validateInput(input: HTMLInputElement | HTMLSelectElement | null) {
    if (!input) return;

    if (input.value && input.value.length < 1) {
      input.setCustomValidity("Value should contain min 1 symbols");
    } else if (!input.value)
      input.setCustomValidity("Value should not be empty");
    else input.setCustomValidity("");
  }

  isFormInvalid = () => {
    return (
      !!this.inputName.current?.validationMessage ||
      !!this.inputFile.current?.validationMessage ||
      !!this.inputDate.current?.validationMessage ||
      !!this.selectRef.current?.validationMessage
    );
  };
  render() {
    const loadImageFile = (event: React.ChangeEvent<HTMLInputElement>) => {
      if (event.target.files && event.target.files[0]) {
        const reader = new FileReader();
        reader.onload = () => {
          if (this.inputFile.current !== null) {
            const imageFile =
              this.inputFile.current.files && this.inputFile.current.files[0];
            if (imageFile !== null) {
              this.setState({
                avatar: URL.createObjectURL(imageFile),
              });
            }
          }
        };
        reader.readAsDataURL(event.target.files[0]);
      }
    };
    return (
      <form
        className={styles.form}
        ref={this.form}
        onSubmit={(e) => {
          e.preventDefault();
          if (this.isFormInvalid()) {
            this.forceUpdate();
          } else {
            this.onSaveUserData(e);
            this.setState({ downloadImg: false });
          }
        }}
        data-testid="react-form"
      >
        <div className={styles.inputBlock}>
          <label htmlFor="" className={styles.label}>
            Name
          </label>
          <input
            type="text"
            ref={this.inputName}
            className={""}
            data-testid="name"
          />
          <p style={{ color: "red" }}>
            {this.inputName.current?.validationMessage}
          </p>
        </div>
        <div className={styles.inputBlock}>
          <label htmlFor="" className={styles.label}>
            Birthday
          </label>
          <input
            type="date"
            ref={this.inputDate}
            className={""}
            data-testid="date"
          />
          <p style={{ color: "red" }}>
            {this.inputDate.current?.validationMessage}
          </p>
        </div>
        <div className={styles.selectBlock}>
          <label htmlFor="" className={styles.label}>
            Nationality
          </label>
          <select ref={this.selectRef} className={""} defaultValue="">
            <option value="" disabled>
              Choose country...
            </option>
            <option value="Russia">Russia</option>
            <option value="Belarus">Belarus</option>
            <option value="Ukraine">Ukraine</option>
            <option value="USA">USA</option>
            <option value="Canada">Canada</option>
          </select>
          <p style={{ color: "red" }}>
            {this.selectRef.current?.validationMessage}
          </p>
        </div>
        <div className={styles.switcherBlock}>
          <span>Female</span>
          <label htmlFor="switcher" className={styles.switcher}>
            <input type="checkbox" id="switcher" ref={this.switcherRef} />
            <span className={styles.slider}></span>
          </label>
          <span>Male</span>
        </div>
        <div className={styles.inputFileBlock}>
          <input
            type="file"
            className={styles.inputFile}
            accept="image/*"
            ref={this.inputFile}
            onChange={(e) => {
              this.setState({ downloadImg: true });
              loadImageFile(e);
            }}
          />
          <label htmlFor="input__file" className={styles.inputFileButton}>
            <span className={styles.inputFileIconWrapper}>
              <img
                className={styles.inputFileIcon}
                src="./img/icon-download.png"
                alt="Выбрать файл"
                width="25"
              />
            </span>
            <span className={styles.inputFileButtonText}>Choose file...</span>
          </label>
          {this.state.downloadImg && (
            <span className={styles.okIcon}>
              <svg
                xmlns="http://www.w3.org/2000/svg"
                viewBox="0 0 48 48"
                width="48px"
                height="48px"
              >
                <path
                  fill="#43A047"
                  d="M40.6 12.1L17 35.7 7.4 26.1 4.6 29 17 41.3 43.4 14.9z"
                />
              </svg>
            </span>
          )}
        </div>
        <div className={styles.agreeBlock}>
          <input
            id="accept"
            type="checkbox"
            className={styles.checkbox}
            ref={this.agreeRef}
          />
          <label htmlFor="accept" className={styles.agree}>
            I agree to the processing of personal data
          </label>
        </div>

        <button className={styles.submitButton} type="submit">
          Save
        </button>
      </form>
    );
  }
}
