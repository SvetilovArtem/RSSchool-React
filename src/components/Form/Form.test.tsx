import { screen, render, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import Form from "./Form";

describe("Form tests", () => {
  it("Form renders", async () => {
    render(<Form setUsers={() => {}} />);
    const element = await screen.findByTestId("react-form");
    expect(element).toBeInTheDocument();
  });
  it("Form submit button renders", () => {
    render(<Form setUsers={() => {}} />);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });
  it("Form submit button renders", () => {
    render(<Form setUsers={() => {}} />);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });
  it("Form: the name must be more than 1 symbol", () => {
    render(<Form setUsers={() => {}} />);
    userEvent.type(screen.getByTestId("name"), "a");
    fireEvent.click(screen.getByRole("button"));
  });
});
