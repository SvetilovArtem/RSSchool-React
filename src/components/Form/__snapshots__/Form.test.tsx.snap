// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form tests Form snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="form"
        data-testid="react-form"
      >
        <div
          class="inputBlock"
        >
          <label
            class="label"
            for=""
          >
            Name
          </label>
          <input
            class=""
            data-testid="name"
            type="text"
          />
          <p
            style="color: red;"
          />
        </div>
        <div
          class="inputBlock"
        >
          <label
            class="label"
            for=""
          >
            Birthday
          </label>
          <input
            class=""
            data-testid="date"
            type="date"
          />
          <p
            style="color: red;"
          />
        </div>
        <div
          class="selectBlock"
        >
          <label
            class="label"
            for=""
          >
            Nationality
          </label>
          <select
            class=""
          >
            <option
              disabled=""
              selected=""
              value=""
            >
              Choose country...
            </option>
            <option
              value="Russia"
            >
              Russia
            </option>
            <option
              value="Belarus"
            >
              Belarus
            </option>
            <option
              value="Ukraine"
            >
              Ukraine
            </option>
            <option
              value="USA"
            >
              USA
            </option>
            <option
              value="Canada"
            >
              Canada
            </option>
          </select>
          <p
            style="color: red;"
          />
        </div>
        <div
          class="switcherBlock"
        >
          <span>
            Female
          </span>
          <label
            class="switcher"
            for="switcher"
          >
            <input
              id="switcher"
              type="checkbox"
            />
            <span
              class="slider"
            />
          </label>
          <span>
            Male
          </span>
        </div>
        <div
          class="inputFileBlock"
        >
          <input
            accept="image/*"
            class="inputFile"
            type="file"
          />
          <label
            class="inputFileButton"
            for="input__file"
          >
            <span
              class="inputFileIconWrapper"
            >
              <img
                alt="Выбрать файл"
                class="inputFileIcon"
                src="./img/icon-download.png"
                width="25"
              />
            </span>
            <span
              class="inputFileButtonText"
            >
              Choose file...
            </span>
          </label>
        </div>
        <div
          class="agreeBlock"
        >
          <input
            class="checkbox"
            id="accept"
            type="checkbox"
          />
          <label
            class="agree"
            for="accept"
          >
            I agree to the processing of personal data
          </label>
        </div>
        <button
          class="submitButton"
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="form"
      data-testid="react-form"
    >
      <div
        class="inputBlock"
      >
        <label
          class="label"
          for=""
        >
          Name
        </label>
        <input
          class=""
          data-testid="name"
          type="text"
        />
        <p
          style="color: red;"
        />
      </div>
      <div
        class="inputBlock"
      >
        <label
          class="label"
          for=""
        >
          Birthday
        </label>
        <input
          class=""
          data-testid="date"
          type="date"
        />
        <p
          style="color: red;"
        />
      </div>
      <div
        class="selectBlock"
      >
        <label
          class="label"
          for=""
        >
          Nationality
        </label>
        <select
          class=""
        >
          <option
            disabled=""
            selected=""
            value=""
          >
            Choose country...
          </option>
          <option
            value="Russia"
          >
            Russia
          </option>
          <option
            value="Belarus"
          >
            Belarus
          </option>
          <option
            value="Ukraine"
          >
            Ukraine
          </option>
          <option
            value="USA"
          >
            USA
          </option>
          <option
            value="Canada"
          >
            Canada
          </option>
        </select>
        <p
          style="color: red;"
        />
      </div>
      <div
        class="switcherBlock"
      >
        <span>
          Female
        </span>
        <label
          class="switcher"
          for="switcher"
        >
          <input
            id="switcher"
            type="checkbox"
          />
          <span
            class="slider"
          />
        </label>
        <span>
          Male
        </span>
      </div>
      <div
        class="inputFileBlock"
      >
        <input
          accept="image/*"
          class="inputFile"
          type="file"
        />
        <label
          class="inputFileButton"
          for="input__file"
        >
          <span
            class="inputFileIconWrapper"
          >
            <img
              alt="Выбрать файл"
              class="inputFileIcon"
              src="./img/icon-download.png"
              width="25"
            />
          </span>
          <span
            class="inputFileButtonText"
          >
            Choose file...
          </span>
        </label>
      </div>
      <div
        class="agreeBlock"
      >
        <input
          class="checkbox"
          id="accept"
          type="checkbox"
        />
        <label
          class="agree"
          for="accept"
        >
          I agree to the processing of personal data
        </label>
      </div>
      <button
        class="submitButton"
        type="submit"
      >
        Save
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
